wafer/compare/admin.py

```diff
@@ -0,0 +1,155 @@
+            date = datetime.date.today() - datetime.timedelta(days=1)
+        elif self.value() == '7 days':
+            date = datetime.date.today() - datetime.timedelta(days=7)
+        elif self.value() == '30 days':
+            date = datetime.date.today() - datetime.timedelta(days=30)
+        if not date:
+            return queryset
+        content_type = ContentType.objects.get_for_model(queryset.model)
+        revisions = Version.objects.filter(content_type=content_type, revision__date_created__gte=date)
+        return queryset.filter(pk__in=[x.object_id for x in revisions])
+
+
+class CompareVersionAdmin(VersionAdmin):
+
+    compare_template = "admin/wafer.compare/compare.html"
+    compare_list_template = "admin/wafer.compare/compare_list.html"
+
+    # Add a compare button next to the History button.
+    change_form_template = "admin/wafer.compare/change_form.html"
+
+    def get_urls(self):
+         urls = super(CompareVersionAdmin, self).get_urls()
+         opts = self.model._meta
+         compare_urls = [
+               url("^([^/]+)/([^/]+)/compare/$", self.admin_site.admin_view(self.compare_view),
+                   name='%s_%s_compare' % (opts.app_label, opts.model_name)),
+               url("^([^/]+)/comparelist/$", self.admin_site.admin_view(self.comparelist_view),
+                   name='%s_%s_comparelist' % (opts.app_label, opts.model_name)),
+         ]
+         return compare_urls + urls
+
+    def compare_view(self, request, object_id, version_id, extra_context=None):
+        """Actually compare two versions."""
+        opts = self.model._meta
+        object_id = unquote(object_id)
+        # get_for_object's ordering means this is always the latest revision.
+        current = self.revision_manager.get_for_object_reference(self.model, object_id)[0]
+        # The reversion we want to compare to
+        revision = self.revision_manager.get_for_object_reference(self.model, object_id).filter(id=version_id)[0]
+        the_diff = []
+        dmp = diff_match_patch()
+
+        for field in current.field_dict:
+            # These exclusions really should be configurable
+            if field == 'id' or field.endswith('_rendered'):
+                continue
+            # KeyError's may happen if the database structure changes
+            # between the creation of revisions. This isn't ideal,
+            # but should not be a fatal error.
+            # Log this?
+            missing_field = False
+            try:
+                cur_val = current.field_dict[field] or ""
+            except KeyError:
+                cur_val = "No such field in latest version"
+                missing_field = True
+            try:
+                old_val = revision.field_dict[field] or ""
+            except KeyError:
+                old_val = "No such field in old version"
+                missing_field = True
+            if missing_field:
+                diffs = dmp.diff_main(old_val, cur_val)
```

Comment:
I imagine this diff could get confusing in cases where one of the values is missing and the other is something that contains words in common with the `No such field in X version` string. Do we perhaps want to construct the diff manually?